core_backend/app/urgency_detection/routers.py

```diff
@@ -0,0 +1,103 @@
+import asyncio
+from typing import Callable
+
+from fastapi import APIRouter, Depends
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from ..auth.dependencies import auth_bearer_token
+from ..database import get_async_session
+from ..llm_call.entailment import detect_urgency
+from ..urgency_rules.models import (
+    get_cosine_distances_from_rules,
+    get_urgency_rules_from_db,
+)
+from ..utils import generate_secret_key
+from .config import (
+    URGENCY_CLASSIFIER,
+    URGENCY_DETECTION_MAX_DISTANCE,
+    URGENCY_DETECTION_MIN_PROBABILITY,
+)
+from .models import save_urgency_query_to_db, save_urgency_response_to_db
+from .schemas import UrgencyQuery, UrgencyResponse
+
+router = APIRouter(
+    dependencies=[Depends(auth_bearer_token)], tags=["Urgency Detection"]
+)
+
+ALL_URGENCY_CLASSIFIERS = {}
+
+
+def urgency_classifier(classifier_func: Callable) -> Callable:
+    """
+    Decorator to register classifier functions
+    """
+    ALL_URGENCY_CLASSIFIERS[classifier_func.__name__] = classifier_func
+    return classifier_func
+
+
+@router.post("/urgency-detect", response_model=UrgencyResponse)
+async def classify_text(
+    urgency_query: UrgencyQuery, asession: AsyncSession = Depends(get_async_session)
+) -> UrgencyResponse:
+    """
+    Classify the urgency of a text message
+    """
+    feedback_secret_key = generate_secret_key()
+    urgency_query_db = await save_urgency_query_to_db(
+        feedback_secret_key, urgency_query, asession
+    )
+
+    classifier = ALL_URGENCY_CLASSIFIERS.get(URGENCY_CLASSIFIER)
+    if not classifier:
+        raise ValueError(f"Invalid urgency classifier: {URGENCY_CLASSIFIER}")
+
+    urgency_response = await classifier(asession, urgency_query)
+
+    await save_urgency_response_to_db(urgency_query_db, urgency_response, asession)
+
+    return urgency_response
+
+
+@urgency_classifier
+async def cosine_distance_classifier(
+    asession: AsyncSession, urgency_query: UrgencyQuery
+) -> UrgencyResponse:
+    """
+    Classify the urgency of a text message using cosine distance
+    """
+
+    cosine_distances = await get_cosine_distances_from_rules(
+        asession, urgency_query.message_text
+    )
+    for _, rule in cosine_distances.items():
+        if float(rule["distance"]) < float(URGENCY_DETECTION_MAX_DISTANCE):
+            return UrgencyResponse(is_urgent=True, details=cosine_distances)
```

Comment:
Shouldn't we return the actual rule it was matched to as well ?

Reply:
It returns all the rules in the details. 
Note that it stops as soon as there is match. So it may match to multiple rules but would returns as soon as it matched one